Add autosave feature

benceruleanlu opened the pull request:
Adds two autosave settings, one for mode ("off (default)" or "after delay") and another for delay in ms (only applies if mode is set to "after delay")

Works how you would expect.

Also, the `status-indicator` in `workflowtab.vue` is hidden even if the workflow is modified and/or not persisted if autosave is set to after delay and the delay value is less than or equal to 3s. Change made for UX purposes, can be debated.

closes #1584

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-3330-Add-autosave-feature-1cd6d73d3650818a9ca1d1f703ef9b78) by [Unito](https://www.unito.io)

ctrlbenlu committed:
Add Auto Save and Auto Save Delay options in settings

ctrlbenlu committed:
Add reasoning for autosave default being off in comment

ctrlbenlu committed:
Add placeholder versionAdded prop

ctrlbenlu committed:
Add botched autosave

ctrlbenlu committed:
Only trigger after graphChanged event

ctrlbenlu committed:
Update status indicator visibility based on Auto Save settings

benceruleanlu committed:
Merge branch 'main' into autosave

ctrlbenlu committed:
nit

ctrlbenlu committed:
Fix bad v-if status indicator logic

ctrlbenlu committed:
Prevent autosave saving when already saving

huchenlei reviewed (changes requested):
Please also add new setting item entry in `apiSchema`.

huchenlei commented on src/composables/useWorkflowPersistence.ts:
```diff
@@ -69,6 +70,67 @@ export function useWorkflowPersistence() {
     }
   }
 
+  const setupAutoSave = () => {
```
For the splitted `useWorkflowAutoSave` composable, can we have some unit tests to verify it works as expected?

benceruleanlu replied:
Sure

ctrlbenlu committed:
Escape autosave functionality into its own composable

ctrlbenlu committed:
Add autosave settings to api schema

ctrlbenlu committed:
Refactor status indicator logic into a computed ref

ctrlbenlu committed:
Prevent frequent settings store access with cached computed refs

benceruleanlu commented on src/composables/useWorkflowPersistence.ts:
```diff
@@ -69,6 +70,67 @@ export function useWorkflowPersistence() {
+      }
+
+      // If autosave is enabled and set to "after delay"
+      if (settingStore.get('Comfy.Workflow.AutoSave') === 'after delay') {
```
```ts
  // Use computed refs to cache autosave settings
  const autoSaveSetting = computed(() =>
    settingStore.get('Comfy.Workflow.AutoSave')
  )
  const autoSaveDelay = computed(() =>
    settingStore.get('Comfy.Workflow.AutoSaveDelay')
  )
  ```

benceruleanlu commented on src/components/topbar/WorkflowTab.vue:
```diff
@@ -12,7 +12,10 @@
+            !workflowOption.workflow.isPersisted) &&
+          (settingStore.get('Comfy.Workflow.AutoSave') === 'off' ||
+            (settingStore.get('Comfy.Workflow.AutoSave') === 'after delay' &&
+              settingStore.get('Comfy.Workflow.AutoSaveDelay') > 3000))
```
```ts
// Use computed refs to cache autosave settings
const autoSaveSetting = computed(() =>
  settingStore.get('Comfy.Workflow.AutoSave')
)
const autoSaveDelay = computed(() =>
  settingStore.get('Comfy.Workflow.AutoSaveDelay')
)

const shouldShowStatusIndicator = computed(() => {
  // Return true if:
  // 1. The shift key is not pressed (hence no override).
  // 2. The workflow is either modified or not yet persisted.
  // 3. AutoSave is either turned off, or set to 'after delay'
  //    with a delay longer than 3000ms.
  return (
    !workspaceStore.shiftDown &&
    (props.workflowOption.workflow.isModified ||
      !props.workflowOption.workflow.isPersisted) &&
    (autoSaveSetting.value === 'off' ||
      (autoSaveSetting.value === 'after delay' && autoSaveDelay.value > 3000))
  )
})
```

ctrlbenlu committed:
Bump versionAdded

ctrlbenlu committed:
Move autosave invokation

ctrlbenlu committed:
Move invokation again...

ctrlbenlu committed:
Add single botched unit test

ctrlbenlu committed:
More unit tests

ctrlbenlu committed:
Refactor auto-save tests to streamline timer management

ctrlbenlu committed:
Catch edge case where post scheduled save is dropped by isSaving

ctrlbenlu committed:
More autosave unit tests

ctrlbenlu committed:
typo

ctrlbenlu committed:
More autosave unit tests

ctrlbenlu committed:
Bump versionAdded

benceruleanlu commented:
Changes since HCL's last review:
1. Moved autosave logic to its own composable, `useWorkflowAutoSave.ts`, invoked in `App.vue`'s `onMounted`
2. Moved setting store gets to computed refs
3. Extracted `WorkflowTab.vue` `status-indicator`'s `v-if` to a computed ref with explanatory comments
4. Fixed edge case where `isSaving` can drop a save that passed scheduling
5. Added unit tests to ensure proper autosaving functionality

ctrlbenlu committed:
Move useWorkflowAutoSave invokation again...

PrometheusDante commented:
Would this feature not cause potential issues as a cached workflow is loaded upon startup, which is often displaying the name of some other workflow?
Either way I often have the title of an entirely different workflow displayed on startup, which upon autosave would actually overwrite the real one. It would be disastrous if that is the case.

benceruleanlu commented:
Hi @PrometheusDante, thanks for raising this concern — that does sound potentially serious, and I want to make sure we fully understand it.

> as a cached workflow is loaded upon startup, which is often displaying the name of some other workflow?

That definitely shouldn't happen, and if it is, it's worth digging into. Could you open a bug report with steps to reproduce this? That will help us track it properly and investigate.

Regarding autosave: it's disabled by default, and it only kicks in after a change is made to the graph. That should mitigate the risk of unintentional overwriting, but we agree this behavior needs to be 100% safe.

We'll keep an eye on this and make sure the autosave logic doesn’t risk overwriting other workflows. Thanks again for the heads up!
